File: evaluation/tasks/onetbb-histogram-local-aggregation/starter/histogram.py
"""Buggy histogram task used by the UXL executable evaluator.

The update below models the common oneTBB mistake of mutating one shared
histogram from parallel workers without a safe reduction pattern.
"""

from __future__ import annotations

from concurrent.futures import ThreadPoolExecutor


def histogram(values: list[int], bins: int = 16, workers: int = 4) -> list[int]:
    counts = [0] * bins

    def update(chunk: list[int]) -> None:
        for value in chunk:
            bucket = value % bins
            current = counts[bucket]
            # A parallel C++ version of this shape would be a data race.
            counts[bucket] = current + 1

    chunks = [values[index::workers] for index in range(workers)]
    with ThreadPoolExecutor(max_workers=workers) as pool:
        list(pool.map(update, chunks))
    return counts


def reference_histogram(values: list[int], bins: int = 16) -> list[int]:
    counts = [0] * bins
    for value in values:
        counts[value % bins] += 1
    return counts


if __name__ == "__main__":
    data = [(index * 17 + 3) % 257 for index in range(1000)]
    print(histogram(data))

File: evaluation/harbor/tasks/onetbb-histogram-local-aggregation/environment/histogram.cpp
#include <cstddef>
#include <vector>

#include <oneapi/tbb/blocked_range.h>
#include <oneapi/tbb/parallel_for.h>

std::vector<std::size_t> histogram(
    const std::vector<int>& values,
    std::size_t bins) {
    std::vector<std::size_t> counts(bins, 0);
    oneapi::tbb::parallel_for(
        oneapi::tbb::blocked_range<std::size_t>(0, values.size()),
        [&](const oneapi::tbb::blocked_range<std::size_t>& range) {
            for (std::size_t index = range.begin(); index != range.end(); ++index) {
                const auto bucket = static_cast<std::size_t>(values[index]) % bins;
                ++counts[bucket];
            }
        });
    return counts;
}
